nilearn/decoding/objective_functions.py

```diff
@@ -0,0 +1,312 @@
+"""
+Common functions and base classes. Used by more specialized modules like
+tv.py, smooth_lasso.py, etc.
+
+"""
+# Author: DOHMATOB Elvis Dopgima,
+#         PIZARRO Gaspar,
+#         VAROQUAUX Gael,
+#         GRAMFORT Alexandre,
+#         PEDREGOSA Fabian
+# License: simplified BSD
+
+from functools import partial
+import numpy as np
+from scipy import linalg
+
+
+def spectral_norm_squared(X):
+    """Computes square of the operator 2-norm (spectral norm) of X
+
+    This corresponds to the Lipschitz constant of the gradient of the
+    squared-loss function:
+
+        w -> .5 * ||y - Xw||^2
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_samples, n_features)
+      Design matrix.
+
+    Returns
+    -------
+    lipschitz_constant : float
+      The square of the spectral norm of X.
+
+    """
+    # On big matrices like those that we have in neuroimaging, svdvals
+    # is faster than a power iteration (even when using arpack's)
+    return linalg.svdvals(X)[0] ** 2
+
+
+def logistic_loss_lipschitz_constant(X):
+    """Compute the Lipschitz constant (upper bound) for the gradient of the
+    logistic sum:
+
+         w -> \sum_i log(1+exp(-y_i*(x_i*w + v)))
+
+    """
+    # N.B: we handle intercept!
+    X = np.hstack((X, np.ones((X.shape[0], 1))))
+    return spectral_norm_squared(X)
+
+
+def squared_loss(X, y, w, compute_energy=True, compute_grad=False):
+    """Compute the MSE error, and optionally, its gradient too.
+
+    The cost / energy function is
+
+        MSE = .5 * ||y - Xw||^2
+
+    A (1 / n_samples) factor is applied to the MSE.
+
+    Parameters
+    ----------
+    X : ndarray, shape (n_samples, n_features)
+        Design matrix.
+
+    y : ndarray, shape (n_samples,)
+        Target / response vector.
+
+    w : ndarray shape (n_features,)
+        Unmasked, ravelized weights map.
+
+    compute_energy : bool, optional (default True)
+        If set then energy is computed, otherwise only gradient is computed.
+
+    compute_grad : bool, optional (default True)
+        If set then gradient is computed, otherwise only energy is computed.
+
+    Returns
+    -------
+    energy : float
+        Energy (returned if `compute_energy` is set).
+
+    gradient : ndarray, shape (n_features,)
+        Gradient of energy (returned if `compute_grad` is set).
+
+    """
+    if not (compute_energy or compute_grad):
+        raise RuntimeError(
+            "At least one of compute_energy or compute_grad must be True.")
+
+    residual = np.dot(X, w) - y
+
+    # compute energy
+    if compute_energy:
+        energy = .5 * np.dot(residual, residual)
+        if not compute_grad:
+            return energy
+
+    grad = np.dot(X.T, residual)
+
+    if not compute_energy:
+        return grad
+
+    return energy, grad
+
+
+def tv_l1_from_gradient(spatial_grad):
```

Comment:
with a leading underscore (these are rather low-level functions, of which users should not be aware)